[FLINK-8981] Add end-to-end test for running on YARN with Kerberos

aljoscha opened the pull request:
This adds a complete Docker container setup and Docker Compose file for
starting a kerberized Hadoop cluster on Docker.

The test script does the following:
 * package "build-target" Flink dist into a tarball
 * build docker container
 * start cluster using docker compose
 * upload tarball and unpack
 * modify flink-conf.yaml to use Kerberos keytab for hadoop-user
 * Run Streaming WordCount Job
 * verify results

We set an exit trap before to ensure that we shut down the docker
compose cluster at the end.

As a prerequisite, this also fixes how we resolve directories in the end-to-end scripts.

aljoscha commented:
This PR adds the test to `flink-ci`: https://github.com/zentol/flink-ci/pull/1

This is a run on my on `flink-ci` fork where the test is run five times without issue: https://travis-ci.org/aljoscha/flink-ci/builds/405995875

zentol commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/README.md:
```diff
@@ -0,0 +1,118 @@
+Run image
+---------
+
+Clone the [Github project](https://github.com/aljoscha/docker-hadoop-secure-cluster) and run
```
point to apache repo instead

aljoscha replied:
fixing

zentol commented on flink-end-to-end-tests/test-scripts/test_yarn_kerberos_docker.sh:
```diff
@@ -0,0 +1,104 @@
+# it's important to run this with higher parallelism, otherwise we might risk that
+# JM and TM are on the same YARN node and that we therefore don't test the keytab shipping
+until docker exec -it master bash -c "export HADOOP_CLASSPATH=\`hadoop classpath\` && /home/hadoop-user/$FLINK_DIRNAME/bin/flink run -m yarn-cluster -yn 3 -ys 1 -ytm 1200 -yjm 800 -p 3 /home/hadoop-user/$FLINK_DIRNAME/examples/streaming/WordCount.jar --output $OUTPUT_PATH"; do
+    echo "Running the Flink job failed, might be that the cluster is not ready yet, retrying ..."
```
is there no way to check whether the cluster is ready? The logs contain several submission failures due to this :/

aljoscha replied:
I'm afraid not, that's why there are the retries around the stuff that deals with HDFS/YARN.

zentol commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/config/log4j.properties:
```diff
@@ -0,0 +1,354 @@
+log4j.logger.org.apache.hadoop.yarn.server.nodemanager=INFO,NODEMANAGER_RFA
+
+## HistoryServer log
+log4j.appender.HISTORYSERVER_RFA=org.apache.log4j.RollingFileAppender
```
do we actually need all this?

aljoscha replied:
most of the stuff in there we don't really need. Removing them

aljoscha replied:
hmm, maybe they help in debugging. Leave them in just in case?

zentol commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+RUN unzip jce_policy-8.zip
+RUN cp /UnlimitedJCEPolicyJDK8/local_policy.jar /UnlimitedJCEPolicyJDK8/US_export_policy.jar $JAVA_HOME/jre/lib/security
+
+ENV HADOOP_VERSION=2.8.4
```
This potentially uses a different hadoop version than the one against flink-dist was built against.

aljoscha replied:
I think the solution in the long run should be to never ship Flink with a Hadoop version, i.e. make the hadoop-free version the default.

zentol replied:
I agree, but for now we still have to ensure that the hadoop version in flink-dist matches, no?

aljoscha replied:
I added a config option to the Dockerfile

aljoscha replied:
and I'm running the nightly tests using the `withoutHadoop` profile

zentol commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+
+ENV HADOOP_VERSION=2.8.4
+
+# ENV HADOOP_URL https://www.apache.org/dist/hadoop/common/hadoop-$HADOOP_VERSION/hadoop-$HADOOP_VERSION.tar.gz
```
remove

aljoscha replied:
removing

zentol commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/bootstrap.sh:
```diff
@@ -0,0 +1,121 @@
+cd $HADOOP_PREFIX/share/hadoop/common ; for cp in ${ACP//,/ }; do  echo == $cp; curl -LO $cp ; done; cd -
+
+# kerberos client
+sed -i "s/EXAMPLE.COM/${KRB_REALM}/g" /etc/krb5.conf
```
`EXAMPLE.COM` is used in several places, is there any way we can set this in a single place? (for example with search&replace if necessary)

aljoscha replied:
`EXAMPLE.COM` is pretty buch the placeholder for this and could be replaced with a different realm in `bootstrap.sh`. But the default is just to still use `EXAMPLE.COM`. I could rename this `TEMPLATE.URL` if you want. 😅

zentol replied:
yeah nvm, I doubt introducing a placeholder really fixes things :/

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+RUN useradd -d /home/hadoop-user -ms /bin/bash -p hadoop-user hadoop-user
+
+# install dev tools
+RUN apt-get update
```
This is a Dockerfile anti-pattern that leads to some cacheing issues:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

aljoscha replied:
fixing

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+RUN cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys
+
+# java
+RUN mkdir -p /usr/java/default && \
```
Can't we use java image as the base image?

aljoscha replied:
I think I possibly could but I don't know exactly what else I then would need to setup to make the whole Hadoop thing work

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+ADD config/container-executor.cfg $HADOOP_HOME/etc/hadoop/container-executor.cfg
+RUN chmod 400 $HADOOP_HOME/etc/hadoop/container-executor.cfg
+RUN chown root:yarn $HADOOP_HOME/etc/hadoop/container-executor.cfg
+# ADD config/log4j.properties $HADOOP_HOME/etc/hadoop/log4j.properties
```
remove?

aljoscha replied:
removing

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+RUN echo "UsePAM no" >> /etc/ssh/sshd_config
+RUN echo "Port 2122" >> /etc/ssh/sshd_config
+
+RUN service ssh start
```
I think it does nothing. Docker does not preserve processes that were run during build.

aljoscha replied:
I don't know myself why this was in there 😅 was in the image I based this on

removing

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+FROM sequenceiq/pam:ubuntu-14.04
+MAINTAINER aljoscha
+
+USER root
```
I think commands in Dockerfile are by default executed as root. So this command is unnecessary.

aljoscha replied:
removing

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/README.md:
```diff
@@ -0,0 +1,118 @@
+
+### Java Keystore
+
+If the Keystroe has been expired, then create a new `keystore.jks`:
```
Keystroe -> Keystore

Won't it be a problem in tests? Will the test start failing one day because of the keystore expired?

aljoscha replied:
fixing the typo but we need the keystore for the SSL setup, which we seem to need for the Kerberos setup

dawidwys replied:
Yes, I rather meant if the expiring of the keystore might be a problem. Could we create the keystore in test?

What is the expiry time for the keystore you use? Maybe setting it to some big number will be enough, but I think the default (365 days) might cause some troubles.

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/bootstrap.sh:
```diff
@@ -0,0 +1,121 @@
+chown mapred:hadoop ${KEYTAB_DIR}/mapred.keytab
+chown yarn:hadoop ${KEYTAB_DIR}/yarn.keytab
+
+service ssh start
```
Can we just make ssh start automatically in Dockerfile?

aljoscha replied:
from a quick search it's not easily possible: https://stackoverflow.com/questions/22886470/start-sshd-automatically-with-docker-container

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/bootstrap.sh:
```diff
@@ -0,0 +1,121 @@
+    nohup sudo -E -u hdfs $HADOOP_PREFIX/bin/hdfs datanode 2>> /var/log/hadoop/datanode.err >> /var/log/hadoop/datanode.out &
+    nohup sudo -E -u yarn $HADOOP_PREFIX/bin/yarn nodemanager 2>> /var/log/hadoop/nodemanager.err >> /var/log/hadoop/nodemanager.out &
+    while true; do sleep 1000; done
+elif [ $1 == "bash" ]; then
```
Is the if necessary?

aljoscha replied:
removing, this was because earlier the setup was meant for more generic/general use cases

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/docker-compose.yml:
```diff
@@ -0,0 +1,87 @@
+      KRB_REALM: EXAMPLE.COM
+      DOMAIN_REALM: kdc.kerberos.com
+
+  slave1:
```
Maybe create just one slave and just use `docker-compose scale`? You run flink from within container anyway. So It could all be uatomatical.

aljoscha replied:
I tried this at the very beginning but this doesn't work because the slaves need well formed hostnames for the Kerberos setup to work (it's tricky with the Kerberos principal names). That's why I did it like this. I also don't like it

dawidwys commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/docker-compose.yml:
```diff
@@ -0,0 +1,87 @@
+    depends_on:
+      - kdc
+    ports:
+      - "50070:50070"
```
I think we do not need to expose ports to the host. We run the flink job from within container anyway.

aljoscha replied:
This was because the setup was meant to be accessible for more generic use and access from outside. I'm removing it.

aljoscha commented:
@zentol I addressed most of your comments. I now added a test in there that verifies the job fails if we don't set a keytab. I'm not running with different Hadoop. It might work but I'm basically setting up a hadoop cluster in docker and I don't know if this is similar enough (or exactly the same, for my purposes) between the versions.

@dawidwys Thanks for the thorough comments, I'll go through them next!

aljoscha commented:
I also ran the new version on `flink-ci`: https://travis-ci.org/aljoscha/flink-ci/builds/406269018

aljoscha commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/docker-compose.yml:
```diff
@@ -0,0 +1,87 @@
+version: '3.5'
+
+networks:
+  docker-hadoop-cluster-network:
```
apparently we don't need it, removing

aljoscha commented on flink-end-to-end-tests/test-scripts/docker-hadoop-secure-cluster/Dockerfile:
```diff
@@ -0,0 +1,159 @@
+USER root
+
+RUN addgroup hadoop
+RUN useradd -d /home/hdfs -ms /bin/bash -G hadoop -p hdfs hdfs
```
will do

aljoscha commented on flink-end-to-end-tests/test-scripts/test_yarn_kerberos_docker.sh:
```diff
@@ -0,0 +1,104 @@
+trap cluster_shutdown INT
+trap cluster_shutdown EXIT
+
+until docker cp $FLINK_TARBALL_DIR/$FLINK_TARBALL master:/home/hadoop-user/; do
```
I did it like this so that rebuilding Flink does not require building the docker image. I know I could do it as one of the last steps but with repeatedly running the test locally I think it's still easier this way. WDYT?

dawidwys replied:
I think if we add it as one of the last steps of the Dockerfile it wouldn't make a difference in build time as all previous layers would be cached anyway. At the same time if we move it to the Dockerfile we will no longer need the loop.

aljoscha commented:
@zentol & @dawidwys I think I addressed all of your comments
